### PR TITLE
Add GetHealthReport commands

### LINKED ISSUE #54200: redfish_facts- Systems - GetHealthReport
##### SUMMARY

This feature should return a JSON structure with the Health status of all the components with Health properties, perhaps grouped by top-level categories like Systems, Chassis, and Manager.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

redfish_facts

##### ADDITIONAL INFORMATION

```yaml

```

### PR DESCRIPTION
**Pull request by billdodd**
##### SUMMARY

Per feature issue #54200, this PR adds `GetHealthReport` commands to the `Systems`, `Chassis`, and `Manager` categories.

The health information is retrieved from the `Status` property in each of the target resources. The target resources for each category are:

Systems: `System`, `Processors`, `Memory`, `SimpleStorage`, `Storage`, `EthernetInterfaces`, `NetworkInterfaces` (`NetworkPorts` and `NetworkDeviceFunctions`) 
Chassis: `Chassis`, `PowerSupplies`, `Fans`, `PCIeDevices`
Manager: `Manager`

Fixes #54200

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

redfish_facts.py
redfish_utils.py

##### ADDITIONAL INFORMATION

Example playbook snippets showing how to invoke the new commands:

```paste below
  - name: Get system health report
    redfish_facts:
      category: Systems
      command: GetHealthReport
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"

  - name: Get chassis health report
    redfish_facts:
      category: Chassis
      command: GetHealthReport
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"

  - name: Get manager health report
    redfish_facts:
      category: Manager
      command: GetHealthReport
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"
```

Example output for the **Systems** category:

```
{
    "ansible_facts": {
        "redfish_facts": {
            "health_report": {
                "entries": [
                    [
                        {   
                            "system_uri": "/redfish/v1/Systems/437XR1138R2"
                        },
                        {
                            "System": {
                                "Status": {
                                    "Health": "OK",
                                    "HealthRollup": "OK",
                                    "State": "Enabled"
                                }       
                            } 
                            "EthernetInterfaces": [
                                {   
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "ethernetinterface_uri": "/redfish/v1/Systems/437XR1138R2/EthernetInterfaces/12446A3B0411"
                                },
                                {   
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "ethernetinterface_uri": "/redfish/v1/Systems/437XR1138R2/EthernetInterfaces/12446A3B8890"
                                }
                            ],
                            "Memory": [
                                {   
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "memory_uri": "/redfish/v1/Systems/437XR1138R2/Memory/DIMM1"
                                },
                                {   
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "memory_uri": "/redfish/v1/Systems/437XR1138R2/Memory/DIMM2"
                                },
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "memory_uri": "/redfish/v1/Systems/437XR1138R2/Memory/DIMM3"
                                },
                                {
                                    "Status": {
                                        "State": "Absent"
                                    },
                                    "memory_uri": "/redfish/v1/Systems/437XR1138R2/Memory/DIMM4"
                                }
                            ],
                            "Processors": [
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },
                                    "processor_uri": "/redfish/v1/Systems/437XR1138R2/Processors/CPU1"
                                },
                                {
                                    "Status": {
                                        "State": "Absent"
                                    },
                                    "processor_uri": "/redfish/v1/Systems/437XR1138R2/Processors/CPU2"
                                }
                            ],
                            "Storage": [
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "HealthRollup": "OK",
                                        "State": "Enabled"
                                    },
                                    "storage_uri": "/redfish/v1/Systems/437XR1138R2/Storage/1"
                                }
                            ],
                        }           
                    ]           
                ],              
                "ret": true         
            }                           
        }                               
    },                              
    "changed": false,               
    "failed": false             
}
```

Example output for the **Chassis** category:

```
{               
    "ansible_facts": { 
        "redfish_facts": {
            "health_report": {
                "entries": [
                    [
                        {
                            "chassis_uri": "/redfish/v1/Chassis/1U"
                        },
                        {
                            "Chassis": {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                }   
                            },  
                            "Fans": [
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },  
                                    "fan_uri": "/redfish/v1/Chassis/1U/Thermal#/Fans/0"
                                },  
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "State": "Enabled"
                                    },  
                                    "fan_uri": "/redfish/v1/Chassis/1U/Thermal#/Fans/1"
                                }   
                            ],  
                            "PCIeDevices": [
                                {
                                    "Status": {
                                        "Health": "OK",
                                        "HealthRollup": "OK",
                                        "State": "Enabled"
                                    },
                                    "pciedevice_uri": "/redfish/v1/Systems/1/PCIeDevices/NIC"
                                }
                            ],
                            "PowerSupplies": [
                                {   
                                    "Status": {
                                        "Health": "Warning",
                                        "State": "Enabled"
                                    },
                                    "powersupply_uri": "/redfish/v1/Chassis/1U/Power#/PowerSupplies/0"
                                }
                            ]       
                        }               
                    ]                   
                ],                  
                "ret": true         
            }                   
        }                       
    },                              
    "changed": false,                   
    "failed": false                     
}
```

Example output for the **Manager** category:

```
{
    "ansible_facts": {
        "redfish_facts": {
            "health_report": {
                "entries": [
                    [
                        {
                            "manager_uri": "/redfish/v1/Managers/Switch1"
                        },
                        {
                            "Manager": {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                }
                            }
                        }
                    ],
                    [
                        {
                            "manager_uri": "/redfish/v1/Managers/Switch2"
                        },
                        {
                            "Manager": {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                }
                            }
                        }
                    ]
                ],
                "ret": true
            }
        }
    },
    "changed": false,
    "failed": false
}
```

### CODE REVIEW

**Comment by billdodd** [comment]
@xmadsen @samerhaj

File: lib/ansible/module_utils/redfish_utils.py [hunk 1]
```diff
@@ -1566,4 +1575,107 @@ def get_system_inventory(self, systems_uri):
...
+                    continue
+            else:  # ex: Memory
+                d = data
+            self.get_health_subsystem(sub, d, health)
+
+        result["entries"] = health
+        return result
+
+    def get_system_health_report(self, systems_uri):
+        subsystems = ['Processors', 'Memory', 'Storage', 'EthernetInterfaces']
```
Comment (mraineri) [inline lib/ansible/module_utils/redfish_utils.py]: Might want to add SimpleStorage and NetworkInterfaces

Comment (billdodd) [reply lib/ansible/module_utils/redfish_utils.py]: Added in new commit.

**Comment by samerhaj** [comment]
Looks good. One nit: Why some of the instance properties are singular (like manager_uri) and the rest are plural (like fans_uri, pciedevices_uri, systems_uri, etc..).

Shouldn't they all be singular (referencing a specific instance, not a collection)?

**Comment by billdodd** [comment]
> Looks good. One nit: Why some of the instance properties are singular (like manager_uri) and the rest are plural (like fans_uri, pciedevices_uri, systems_uri, etc..).
> 
> Shouldn't they all be singular (referencing a specific instance, not a collection)?

Good catch. I updated to make those URI keys singular.

**Comment by mraineri** [comment]
shipit

**Comment by tomasg2012** [comment]
shipit

**Comment by samerhaj** [comment]
shipit

**Comment by billdodd** [comment]
shipit

**Comment by gundalow** [comment]
@samerhaj @tomasg2012 Hi, could you please re-review this and add your `s h i p i t` if you are happy, the reviews get reset after code has been pushed. Thanks in advance!

**Comment by tomasg2012** [comment]
shipit

**Comment by gundalow** [comment]
Rebate please. I wonder how we can avoid the problem of conflicts in the future?

**Comment by tomasg2012** [comment]
Example using command
` % ansible-playbook -i inventory.yml {...}/get_chassis_health.yml`
```
PLAY [Chassis Health Report] *********************************************************************************************

TASK [Define output file] ************************************************************************************************
included: /home/tomp/redfish/Redfish-Ansible-Playbooks/playbooks/chassis/create_output_file.yml for idrac

TASK [Define timestamp] **************************************************************************************************
ok: [idrac]

TASK [Define file to place results] **************************************************************************************
ok: [idrac]

TASK [Create dropoff directory for host] *********************************************************************************
[DEPRECATION WARNING]: Distribution Ubuntu 18.04 on host idrac should use /usr/bin/python3, but is using /usr/bin/python 
for backward compatibility with prior Ansible releases. A future Ansible release will default to using the discovered 
platform python for this host. See https://docs.ansible.com/ansible/devel/reference_appendices/interpreter_discovery.html
 for more information. This feature will be removed in version 2.12. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [idrac]

TASK [Getting chassis health report] *************************************************************************************
ok: [idrac]

TASK [Copy results to output file] ***************************************************************************************
changed: [idrac]

PLAY RECAP ***************************************************************************************************************
idrac                      : ok=6    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

....
{
    "changed": false,
    "failed": false,
    "redfish_facts": {
        "health_report": {
            "entries": [
                [
                    {
                        "chassis_uri": "/redfish/v1/Chassis/System.Embedded.1"
                    },
                    {
                        "Chassis": {
                            "Status": {
                                "Health": "Critical",
                                "HealthRollup": "Critical",
                                "State": "Enabled"
                            }
                        },
                        "Fans": [
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.1A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.2A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.3A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.4A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.5A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.6A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.7A"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.1B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.2B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.3B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.4B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.5B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.6B"
                            },
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "fan_uri": "/redfish/v1/Chassis/System.Embedded.1/Sensors/Fans/0x17%7C%7CFan.Embedded.7B"
                            }
                        ],
                        "PowerSupplies": [
                            {
                                "Status": {
                                    "Health": "OK",
                                    "State": "Enabled"
                                },
                                "powersupply_uri": "/redfish/v1/Chassis/System.Embedded.1/Power/PowerSupplies/PSU.Slot.1"
                            },
                            {
                                "Status": {
                                    "Health": "Critical",
                                    "State": null
                                },
                                "powersupply_uri": "/redfish/v1/Chassis/System.Embedded.1/Power/PowerSupplies/PSU.Slot.2"
                            }
                        ]
                    }
                ],
                [
                    {
                        "chassis_uri": "/redfish/v1/Chassis/Enclosure.Internal.0-1:RAID.Integrated.1-1"
                    },
                    {
                        "Chassis": {
                            "Status": {
                                "Health": "OK",
                                "HealthRollup": "OK",
                                "State": "Enabled"
                            }
                        }
                    }
                ]
            ],
            "ret": true
        }
    }
}
```

**Comment by tomasg2012** [comment]
shipit

**Comment by mraineri** [comment]
shipit

**Comment by billdodd** [comment]
> Rebate please. I wonder how we can avoid the problem of conflicts in the future?

@gundalow Rebased, re-reviewed and ready to merge.

I suspect there may be some refactoring or reorganizing work that could be done that would help reduce the number of merge conflicts. But I don't have any concrete ideas at this time. Though I am hoping that as the code base gets more mature and feature set more complete the merge conflicts will naturally die down to some degree.

**Comment by gundalow** [comment]
@billdodd Yup, I think that's a fair comment. If their is anything you think I can help with then please do shout out.

Everybody: Excellent work :)

**Pull request merged**